woocommerce/woocommerce-ios
# Edit Product: add product details bottom sheet

Add product details bottom sheet for #2053 

- [x] ⚠️ Update PR base to `develop` after https://github.com/woocommerce/woocommerce-ios/pull/2210 is merged ⚠️ 

This PR presents the more details bottom sheet with product categories that aren't set yet with the following logic so that each product category is either shown in the product form or the bottom sheet except for the price (always visible in the product form).

product category | when it's set (shown in the list view, not in the more details bottom sheet)
-- | --
inventory | at least one the following is set: sku, stock quantity
shipping | at least one the following is set: weight, length, width, height
categories | product has at least one category
short description | product's short description is not empty


## Changes

- Added an extension helper to `Optional` where the type is `String` to determine whether the string is nil or empty
- Created `ProductFormBottomSheetAction` for each action that could be in the product form bottom sheet, along with the title and subtitle displayed in the bottom sheet row
  - Added an extension to determine whether the action is visible based on the product
  - Similarly, added an extension to determine whether a `ProductFormSection.SettingsRow` is visible based on the product
- Created `ProductFormBottomSheetListSelectorCommand` for rendering the bottom sheet list selector
- In `ProductFormViewController`, handled each bottom sheet selection by navigating to the edit screen based on the action; updated the more details CTA visibility when the product is updated
- Added a `Product` extension to determine whether shipping is enabled, and updated another existing use case in `DefaultProductFormTableViewModel` where we determine whether to show the shipping settings row
- Added unit tests

## Testing

- Go to the Products tab
- Tap on a simple Product --> should see a "+ Add more details" CTA at the bottom; should only see rows that have non-empty data (as at the top of the PR description)
- Tap on the "+ Add more details" CTA at the bottom --> should see a list selector with actions to add more details that are missing
    - If the product is virtual or downloadable, the shipping row shouldn't be visible
- Tap on a row --> it should navigate to the corresponding edit screen
- Navigate back to the product form and repeat the above for each row

## Example screencast/screenshots

![simulator](https://user-images.githubusercontent.com/1945542/81069799-f887f900-8f14-11ea-87f7-d672feb3f3b2.gif)

Update release notes:
